File: clients/android/app/src/main/java/com/remoteeverything/core/api/CatalogOutcome.kt
package com.remoteeverything.core.api

import com.remoteeverything.core.model.AppInfo
import com.remoteeverything.core.model.ClientError
import com.remoteeverything.core.model.ErrorCode

/**
 * What one catalog answer means for the screen that shows it. The wire says
 * "this machine is off" as an answer, not a refusal, so the screen it leads to
 * is a state rather than an error; only a device that is no longer allowed, or a
 * gateway that cannot be read, are the other ways a node's screen can end up.
 */
sealed interface CatalogOutcome {
    data class Apps(val apps: List<AppInfo>) : CatalogOutcome
    data object Offline : CatalogOutcome
    data object Unauthorized : CatalogOutcome
    data object GatewayTrouble : CatalogOutcome

    companion object {
        fun forAnswer(answer: CatalogResponse): CatalogOutcome =
            if (!answer.computer_connected || answer.code == CatalogCode.COMPUTER_OFFLINE) {
                Offline
            } else {
                Apps(answer.apps.map { it.toAppInfo() })
            }

        fun forRefusal(error: ClientError): CatalogOutcome = when (error.code) {
            ErrorCode.UNAUTHORIZED, ErrorCode.NODE_NOT_FOUND, ErrorCode.NODE_REQUIRED -> Unauthorized
            ErrorCode.COMPUTER_OFFLINE -> Offline
            else -> GatewayTrouble
        }
    }
}

File: clients/ios/RemoteEverything/Core/API/CatalogOutcome.swift
import Foundation

/// What a catalog answer means for the screen — the same four conclusions every
/// client draws, so the behaviour fixtures can assert them by name.
enum CatalogOutcome: Equatable {
    case apps([AppInfo])
    case offline
    case unauthorized
    case gatewayTrouble

    /// The screen one catalog answer calls for. `computer_offline` is an answer,
    /// not a failure: the machine is off, and the screen says so.
    static func forAnswer(_ response: CatalogResponse) -> CatalogOutcome {
        guard response.computerConnected else { return .offline }
        return .apps(response.applicationInfos)
    }

    /// The screen one refusal calls for.
    static func forRefusal(_ error: ClientError) -> CatalogOutcome {
        switch error.code {
        case .unauthorized, .nodeRequired, .nodeNotFound:
            return .unauthorized
        case .computerOffline:
            return .offline
        default:
            return .gatewayTrouble
        }
    }
}
